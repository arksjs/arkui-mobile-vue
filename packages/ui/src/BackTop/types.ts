import type { OnClick } from '../helpers'

export interface BackTopProps {
  visibleHeight?: number | string
  animated?: boolean
  offset?: number | number[]
  enableSafeAreaInsets?: boolean
}

export interface BackTopEmits {
  onClick?: OnClick
}
