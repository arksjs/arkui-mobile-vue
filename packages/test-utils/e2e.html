<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
