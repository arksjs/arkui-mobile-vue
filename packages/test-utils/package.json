{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
