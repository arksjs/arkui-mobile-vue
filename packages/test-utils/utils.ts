export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
