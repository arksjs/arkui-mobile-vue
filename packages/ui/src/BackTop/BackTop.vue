<template>
  <button class="ak-back-top" :style="styles" @click="onClick">
    <slot>
      <Icon :icon="UpCircleOutlined" />
    </slot>
  </button>
</template>

<script lang="ts">
import { defineComponent, computed, toRef, ref, onMounted } from 'vue'
import type { PropType } from 'vue'
import { Icon } from '../Icon'
import { getScrollTop, scrollTo } from '../helpers/dom'
import { useSafeAreaInsets } from '../hooks/use-safe-area-insets'
import { useScroll } from '../hooks/use-scroll'
import { emitEventValidator } from '../helpers/validator'
import UpCircleOutlined from '../Icon/icons/UpCircleOutlined'
import type { OnClick, PropsToEmits } from '../helpers/types'
import { DEFAULT_VISIBLE_HEIGHT, getStyles } from './util'
import { getNumber } from '../helpers/util'
import type { BackTopEmits } from './types'

export default defineComponent({
  name: 'ak-back-top',
  components: { Icon },
  props: {
    visibleHeight: {
      type: [Number, String],
      default: DEFAULT_VISIBLE_HEIGHT
    },
    animated: {
      type: Boolean,
      default: true
    },
    offset: {
      type: [Number, Array] as PropType<number | number[]>,
      default: 0
    },
    enableSafeAreaInsets: {
      type: Boolean,
      default: true
    }
  },
  emits: {
    click: emitEventValidator
  } as PropsToEmits<BackTopEmits>,
  setup(props, { emit }) {
    const isShow = ref(false)
    const docEl = ref<HTMLElement>()

    function toTop() {
      scrollTo(document, 0, props.animated)
    }

    const onClick: OnClick = e => {
      toTop()

      emit('click', e)
    }

    const { safeAreaInsets } = useSafeAreaInsets(
      toRef(props, 'enableSafeAreaInsets')
    )

    const styles = computed(() =>
      getStyles(props.offset, isShow.value, safeAreaInsets)
    )

    function updateShow() {
      isShow.value =
        getScrollTop() >= getNumber(props.visibleHeight, DEFAULT_VISIBLE_HEIGHT)
    }

    useScroll(docEl, updateShow)

    onMounted(() => {
      updateShow()

      docEl.value = document.documentElement
    })

    return {
      toTop,
      onClick,
      isShow,
      styles,
      UpCircleOutlined
    }
  }
})
</script>
