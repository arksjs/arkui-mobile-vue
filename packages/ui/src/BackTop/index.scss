@import '../style/var.scss';

.ta-back-top {
  --ta-color: #{$title-color};

  position: fixed;
  right: 20px;
  bottom: 20px;
  z-index: 1000;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  min-width: 40px;
  min-height: 40px;
  font-size: 17px;
  color: var(--ta-color);
  background: none;
  border: none;
  padding: 0;
  margin: 0;
  user-select: none;
  outline: 0;

  .ta-icon {
    --ta-icon-size: 40px;
    --ta-icon-color: var(--ta-color);

    display: block;
  }
}
