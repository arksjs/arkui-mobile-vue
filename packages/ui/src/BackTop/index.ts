import { withInstall } from '../helpers/with-install'
import BackTop from './BackTop.vue'

export { BackTop }
export const install = withInstall(BackTop)
export default BackTop
