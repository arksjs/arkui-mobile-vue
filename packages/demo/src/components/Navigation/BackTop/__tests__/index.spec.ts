import { shallowMount, mount } from '@vue/test-utils'
import { BackTop } from '@/BackTop'

describe('BackTop', () => {
  test('should render default correctly', () => {
    const wrapper = shallowMount(BackTop, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop offset=[0, -50] correctly', () => {
    const wrapper = shallowMount(BackTop, {
      props: {
        offset: [0, -50]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render set default slot correctly', () => {
    const wrapper = mount(BackTop, {
      slots: {
        default: 'BackTop'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('test visible by scroll', async () => {
    const wrapper = mount(BackTop, {
      attachTo: document.body
    })

    expect(wrapper.isVisible()).toBeFalsy()
  })
})
