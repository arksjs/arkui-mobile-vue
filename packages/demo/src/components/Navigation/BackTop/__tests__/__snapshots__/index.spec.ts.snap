// Vitest Snapshot v1

exports[`BackTop > should render default correctly 1`] = `
"<button class=\\"fx-back-top\\" style=\\"transform: translate3d(0px, 0px, 0px); display: none;\\">
  <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\"></icon-stub>
</button>"
`;

exports[`BackTop > should render prop offset=[0, -50] correctly 1`] = `
"<button class=\\"fx-back-top\\" style=\\"transform: translate3d(0px, -50px, 0px); display: none;\\">
  <icon-stub icon=\\"[object Object]\\" spin=\\"false\\" size=\\"0\\" width=\\"0\\" height=\\"0\\"></icon-stub>
</button>"
`;

exports[`BackTop > should render set default slot correctly 1`] = `"<button class=\\"fx-back-top\\" style=\\"transform: translate3d(0px, 0px, 0px); display: none;\\">BackTop</button>"`;
